tsdoc/src/parser/NodeParser.ts

```diff
@@ -1826,105 +1821,268 @@ export class NodeParser {
+
+        // Recurse into the child element to generate all nested child nodes.
+        const childNode: DocNode = this._parseXmlElement(tokenReader);
+
+        // Stop at the first error.
+        if (childNode.kind === DocNodeKind.ErrorText) {
+          return childNode;
+        }
+
+        // Add the child node to the list of child nodes.
+        childNodes.push(childNode);
+        continue;
+      }
+
+      // We've hit an XML text node.
+
+      const childNode: ResultOrFailure<DocNode> = this._parseXmlInnerText(tokenReader);
+
+      if (isFailure(childNode)) {
+        return this._backtrackAndCreateErrorRangeForFailure(
+          tokenReader,
+          startMarker,
+          startTagClosingDelimiterMarker,
+          'Invalid XML element: ',
+          childNode
+        );
+      }
+
+      childNodes.push(childNode);
+    }
+
+    const endMarker: number = tokenReader.createMarker();
+
+    const endTagLessThanToken: Token = tokenReader.peekToken();
+    if (endTagLessThanToken.kind !== TokenKind.LessThan) {
+      return this._backtrackAndCreateError(
         tokenReader,
-        marker,
-        'The HTML tag has invalid syntax: ',
-        failure
+        endMarker,
+        TSDocMessageId.MissingXmlEndTag,
+        'Expecting a closing tag starting with "</"'
       );
     }
     tokenReader.readToken();
 
-    const closingDelimiterExcerpt: TokenSequence = tokenReader.extractAccumulatedSequence();
+    // Read the preceding "/" from the closing tag
+    const slashToken: Token = tokenReader.peekToken();
+    if (slashToken.kind !== TokenKind.Slash) {
+      return this._backtrackAndCreateError(
+        tokenReader,
+        endMarker,
+        TSDocMessageId.MissingXmlEndTag,
+        'Expecting an XML tag starting with "</"'
+      );
+    }
+
+    tokenReader.readToken();
+
+    const endTagOpeningDelimiterExcerpt: TokenSequence = tokenReader.extractAccumulatedSequence();
+
+    const endTagNameStartMarker: number = tokenReader.createMarker();
+    const endTagNameExcerpt: ResultOrFailure<TokenSequence> = this._parseXmlName(tokenReader);
+    const endTagnameEndMarker: number = tokenReader.createMarker();
+    if (isFailure(endTagNameExcerpt)) {
+      return this._backtrackAndCreateErrorForFailure(
+        tokenReader,
+        endMarker,
+        'Invalid XML end tag name: ',
+        endTagNameExcerpt
+      );
+    }
 
-    // NOTE: We don't read excerptParameters.separator here, since if there is any it
-    // will be represented as DocPlainText.
+    // Verify that the tag names are matching, if they aren't create a failure.
+    if (endTagNameExcerpt.toString() !== startTagNameExcerpt.toString()) {
+      return this._backtrackAndCreateErrorRange(
+        tokenReader,
+        endTagNameStartMarker,
+        endTagnameEndMarker,
+        TSDocMessageId.XmlTagNameMismatch,
+        `Expecting closing tag name "${endTagNameExcerpt.toString()}" to match opening tag name "${startTagNameExcerpt.toString()}"`
```

Comment:
In the TSDoc Playground, the representation of this error is not quite right:

![image](https://user-images.githubusercontent.com/4673363/183865943-4b429a76-2fbd-4627-ab93-616f71109abd.png)

For the purposes of syntax highlighting and roundtripping, the parser tries to ensure that every input character corresponds to some AST excerpt, without any characters getting discarded.  

Some of the unit tests validate this property:
https://github.com/microsoft/tsdoc/blob/d8ce4aedcf64b5c948e7152ba02ac6c7bdedc9be/tsdoc/src/parser/__tests__/TestHelpers.ts#L177-L180

It raises a design question of how best to represent `</c>` in the AST?  

1. If it's too small of a fragment to be meaningful, then we could represent it as `ErrorText="</c>"`. 
2. For a richer representation, we could represent it as a loose `XmlElement` who is missing its opening excerpts. And maybe `<b>` would become a loose `XmlElement` who is missing its closing excerpts. (`<a>` does NOT need to be a child of `<b>` in this representation.)
3. An "intelligent" possibility would be to interpret `</c>` as the closing tag for `<b>`.  So `<b>` is a normal nesting element, just accompanied by an error message.

I don't have a strong opinion, but I suspect that 3 is maybe overengineering, and the decision of 1-vs-2 depends on how good we want the syntax highlighting to be, and how much time we have to spend on this PR. 🙂

Reply:
I'll try out option 2 and see if it's feasible, it seems like the most ideal solution in my opinion.

Reply:
@octogonz I'm re-reading option 3 and just to be clear what you're proposing for that option is `ErrorText=<b>`, instead of setting the error text to `</c>`? If so, I think that may be easier than option 2.

Reply:
To be clear, the parser has two ways of modeling syntax errors:
1. `DocErrorText` which will get syntax highlighted as red characters, plus an accompanying error message in the logs
2. a normal AST node such as `DocXmlElement`, whose properties are perhaps only partially initialized, plus an accompanying error message in the logs

The advantage of 2 is that syntax highlighting can be applied to the recognized characters, rather than having the entire span be red text.  The disadvantage of 2 is that if the AST node is too degenerate, then it is awkward for consumers of the API to handle, or perhaps completely incorrect (for example if we interpreted some stray characters `<y` as a `DocXmlElement`).

> just to be clear what you're proposing for that option is `ErrorText=<b>` instead of setting the error text to `</c>`? If so, I think that may be easier than option 2.

No, what I meant for option 3 was that excerpts `<b>` and `</c>` would belong to the same `DocXmlElement`.  I think it's not easier, because the some judgement calls are needed for less obvious such as `<a><b><c></c></a></b>` or `<a><b><c></c></a>`.

(If you want to save time, for an input like `<b><a></a></c>`, the least effort might be to generate `<b>` and `</c>` as `DocErrorText` -- it's not elegant, but we could always improve it later.)

I don't have a strong preference -- it's up to you.

Reply:
Hi again @octogonz,

I actually went with option 3. The main reason I did this was because of the behaviors I observed when trying different XML parsers. Most of them properly figure out what's *supposed* to be the end tag for a given start tag. As for cases where the error non-obvious, I followed the errors I saw in the other parsers.

For example, 
```xml
<a><b><c></c></a></b>
```

Would yield the following AST:
```
	- XmlElement
        * XmlStartTag_OpeningDelimiter="<"
        * XmlElement_Name="a"
        * XmlStartTag_ClosingDelimiter=">"
        - XmlElement
          * XmlStartTag_OpeningDelimiter="<"
          * XmlElement_Name="b"
          * XmlStartTag_ClosingDelimiter=">"
          - XmlElement
            * XmlStartTag_OpeningDelimiter="<"
            * XmlElement_Name="c"
            * XmlStartTag_ClosingDelimiter=">"
            * XmlEndTag_OpeningDelimiter="</"
            * XmlEndTag_Name="c"
            * XmlEndTag_ClosingDelimiter=">"
          * XmlEndTag_OpeningDelimiter="</"
          * XmlEndTag_Name="a"
          * XmlEndTag_ClosingDelimiter=">"
        * XmlEndTag_OpeningDelimiter="</"
        * XmlEndTag_Name="b"
        * XmlEndTag_ClosingDelimiter=">"
```

Along with the following error in the log:
`Expected closing tag "<b/>" but received "<a/>" instead.`

A loose start tag will always try to find the next available *unmatched* end tag. If the end tag doesn't match a log error will be emitted and the AST for that particular element would be the start tag and then mismatched end tag as its end tag element.

`<a><b><c></c></a>`

Here `c` is a valid XML element, however, `b` is not because the next closest end tag is `</a>`. `a` is also not a valid XML element because the next closing tag closing tag `</c>` is matched with `<c>` already, the next closing tag after `</c>` is matched with `b`. This results in `a` not having a closing tag.

It can represented visually like so:
<img width="188" alt="Untitled" src="https://user-images.githubusercontent.com/77477100/186960473-beddbc7b-9248-4253-a0c7-a2ab82d36916.png">


Cases where the element is not large enough to create meaningful `DomXmlElement`, the text is simply considered `DocErrorText`.

With this I've been able to get mostly all of the valid syntax highlighting for loose nodes to work in various situations. I look forward to hearing your thoughts and feedback on this.